bmSettings.lua

```diff
@@ -34,6 +38,369 @@ BMSettings.Constants.UpdateRates = {
+            button_number INTEGER NOT NULL,
+            button_id TEXT NOT NULL
+        );
+        CREATE TABLE IF NOT EXISTS buttons (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            button_number TEXT NOT NULL,
+            button_label TEXT NOT NULL,
+            button_render INTEGER NOT NULL,
+            button_text_color TEXT,
+            button_button_color TEXT,
+            button_cached_countdown INTEGER,
+            button_cached_cooldown INTEGER,
+            button_cached_toggle_locked INTEGER,
+            button_cached_last_run NUMERIC,
+            button_label_mid_x INTEGER,
+            button_label_mid_y INTEGER,
+            button_cached_label TEXT,
+            button_cmd TEXT,
+            button_evaluate_label INTEGER,
+            button_show_label INTEGER,
+            button_icon INTEGER,
+            button_icon_type TEXT,
+            button_icon_lua TEXT,
+            button_timer_type TEXT,
+            button_cooldown TEXT
+        );
+        CREATE TABLE IF NOT EXISTS windows (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            server TEXT NOT NULL,
+            character TEXT NOT NULL,
+            window_id INTEGER NOT NULL,
+            window_fps INTEGER NOT NULL,
+            window_button_size INTEGER NOT NULL,
+            window_advtooltip INTEGER NOT NULL,
+            window_compact INTEGER NOT NULL,
+            window_hide_title INTEGER NOT NULL,
+            window_width INTEGER NOT NULL,
+            window_height INTEGER NOT NULL,
+            window_x INTEGER NOT NULL,
+            window_y INTEGER NOT NULL,
+            window_visible INTEGER NOT NULL,
+            window_font_size INTEGER NOT NULL,
+            window_locked INTEGER NOT NULL,
+            window_theme TEXT NOT NULL,
+            window_set_id INTEGER NOT NULL,
+            window_set_name TEXT NOT NULL
+        );
+        CREATE TABLE IF NOT EXISTS characters (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            character TEXT NOT NULL,
+            character_locked INTEGER NOT NULL,
+            character_hide_title INTEGER NOT NULL
+        );
+    ]])
+    return db
+end
+
+function BMSettings:saveToDB(db, query, ...)
+    local stmt = db:prepare(query)
+    stmt:bind_values(...)
+    stmt:step()
+    stmt:finalize()
+end
+
+function BMSettings:loadFromDB(db, query, ...)
+    local stmt = db:prepare(query)
+    stmt:bind_values(...)
+    local data = {}
+    for row in stmt:nrows() do
+        table.insert(data, row)
+    end
+    stmt:finalize()
+    return data
+end
+
+function BMSettings:updateButtonDB(buttonData, id)
+    local db = self:InitializeDB()
+    local icon = tonumber(buttonData.Icon) or 0
+    local showLabel = icon == 0 and true or buttonData.ShowLabel ~= nil and buttonData.ShowLabel or true
+    self:saveToDB(db, [[
+    INSERT OR REPLACE INTO buttons (
+        button_number, button_label, button_render, button_text_color, button_button_color,

```

Comment:
None of these insert or replace statements will work because they will only replace if the key collides but the only key you provide to the table is the id auto increment which is not specified here.  See:
SQLite Insert or Replace
SQLite’s INSERT OR REPLACE statement is an alias for the INSERT statement with the OR clause, which specifies the action to take when a row with the same primary key already exists. When a conflict occurs, the statement deletes the existing row and inserts a new one.